wagtail/permission_policies/collections.py

```diff
@@ -22,47 +47,34 @@ def _check_perm(self, user, actions, collection=None):
         if user.is_superuser:
             return True
 
-        collection_permissions = GroupCollectionPermission.objects.filter(
-            group__user=user,
-            permission__in=self._get_permission_objects_for_actions(actions),
+        collection_permissions = self._get_user_permission_objects_for_actions(
+            user, actions
         )
 
         if collection:
-            collection_permissions = collection_permissions.filter(
-                collection__in=collection.get_ancestors(inclusive=True)
-            )
+            collection_permissions = {
+                permission
+                for permission in collection_permissions
+                if collection.is_descendant_of(permission.collection)
+                or collection.pk == permission.collection_id
+            }
 
-        return collection_permissions.exists()
+        return bool(collection_permissions)
 
     def _collections_with_perm(self, user, actions):
         """
         Return a queryset of collections on which this user has a GroupCollectionPermission
         record for any of the given actions, either on the collection itself or an ancestor
         """
-        # Get the permission objects corresponding to these actions
-        permissions = self._get_permission_objects_for_actions(actions)
-
-        # Get the collections that have a GroupCollectionPermission record
-        # for any of these permissions and any of the user's groups;
-        # create a list of their paths
-        collection_root_paths = Collection.objects.filter(
-            group_permissions__group__in=user.groups.all(),
-            group_permissions__permission__in=permissions,
-        ).values_list("path", flat=True)
+        permissions = self._get_user_permission_objects_for_actions(user, actions)
 
-        if collection_root_paths:
-            # build a filter expression that will filter our model to just those
-            # instances in collections with a path that starts with one of the above
-            collection_path_filter = Q(path__startswith=collection_root_paths[0])
-            for path in collection_root_paths[1:]:
-                collection_path_filter = collection_path_filter | Q(
-                    path__startswith=path
-                )
+        collections = Collection.objects.none()
+        for perm in permissions:
+            collections |= Collection.objects.descendant_of(
```

Comment:
note to self: this is equivalent to
`Q(path__startswith=perm.collection.path) & Q(depth__gte=perm.collection.depth) & ~Q(pk=perm.collection.pk)`

Reply:
`~Q(pk=perm.collection.pk)` does not apply if `inclusive=True` though

Reply:
yup. had an unsaved edit, and got distracted